nixd/lspserver/src/Connection.cpp

```diff
@@ -222,24 +223,53 @@ bool InboundPort::readStandardMessage(std::string &JSONString) {
 bool InboundPort::readDelimitedMessage(std::string &JSONString) {
   JSONString.clear();
   llvm::SmallString<128> Line;
-  bool IsInputBlock = false;
+  bool IsJSONBlock = false;
```

Comment:
If we do need to implement this functionality in `readStandardMessage`, I suggest using a **tagged union** (also known as a discriminated union) to achieve it. The purpose of doing this is to avoid the situation where both `IsJSONBlock` and `NixDocURI` are `true` at the same time.

Here's an example of how we could structure it:

```cpp
struct TestMessageState {
    enum class ContextKind { JSONBlock, NixBlock /*, ... */ };

    ContextKind kind;
};
```

Reply:
Good point on parser states being better modeled by an enum than two booleans. I refactored the code accordingly :)

Sidenote: I didn't introduce a tagged union to also contain the NixDocURI because it seems like tagged unions in C++ aren't as ergonomic as the algebraic data types in Rust (just called enums there).